### PR TITLE
Modify composite function attributes API

### PR DESCRIPTION
**Pull request by StephenSmith25**
**Description of work.**
This PR modifies the API of attributes in Composite function, so it is similar to the API for setting/getting parameters of a composite function. This is to avoid having to write different code for manipulating attributes on composite functions and IFunctions.
Fundamentally, the changes can be seen in the script, where all the attributes are now printed, with an appropriate prefix.
```
func = FunctionFactory.Instance().createInitialized("name=GausOsc,A=0.2,Sigma=0.2,Frequency=0.1,Phi=0;name=TeixeiraWaterSQE,Q=8.9884656743115785e+307,WorkspaceIndex=2147483647,Height=1,DiffCoeff=2.3,Tau=1.25,Centre=0")
print("Composite attributes", func.attributeNames())
```

**To test:**
1. Code review
2. Generally these changes should be covered by unittests, but testing the following fitting interfaces would be a good place to start:
3. Test the workbench fitting browser by loading data and adding an appropriate fit function - perform a fit.
4. Test the Muon analysis fitting browser: Open the Muon Analysis GUI -> Load any data -> Go to fitting -> Add and remove some fit functions -> Run the fit
5. Test indirect: Open the Indirect Data Analayis GUI -> ConvFit -> Load data -> Add fit functions, e.g 2 Lorentzians  -> Click the full function view tickbox -> Run the fit

 

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/widgets/common/test/ConvolutionFunctionModelTest.h [hunk 1]
```diff
@@ -361,11 +361,11 @@ class ConvolutionFunctionModelTest : public CxxTest::TestSuite {
     TS_ASSERT_EQUALS(
         fitFunctionAsString,
         "composite=MultiDomainFunction,NumDeriv=true;(composite=Convolution,"
-        "FixResolution=true,NumDeriv=true,$domains=i;name=Resolution,Workspace="
+        "NumDeriv=true,FixResolution=true,$domains=i;name=Resolution,Workspace="
```
Comment (martyngigg) [inline qt/widgets/common/test/ConvolutionFunctionModelTest.h:364]: If a string existed from the current implementation I assume the order here wouldn't matter and the generator can handle it?

**Review by martyngigg — CHANGES_REQUESTED** [review]
This looks very nice. It will be good to be able to treat the Composite as a single function as was intended. A few comments/questions inline.

As a general comment/question - what do you think of marking all of the `const` getters `[[nodiscard]]`?

**Comment by StephenSmith25** [comment]
> This looks very nice. It will be good to be able to treat the Composite as a single function as was intended. A few comments/questions inline.
> 
> As a general comment/question - what do you think of marking all of the `const` getters `[[nodiscard]]`?

I was wondering about using `[[nodiscard]]`. The textbook example in the STL appears to the empty() method, which when marked as `[[nodiscard]]` can avoid mistakes such as:

`vec.empty() # this should actually be vec.clear()`

`.empty` is a good example as its not clear—through the function name alone—what the method does. Another example is marking `std::allocate` as `[[nodiscard]]` which avoids us throwing away the results from costly computations. 

I think the decisions we have here is how liberal to be with the attribute, as it is clear that a function named `getParameter` returns something, but if we do add `[[nodiscard]]` to the signature at least it avoids someone calling the method unnecessarily (possibly by mistake) as they always have to assign the result to something.

Personally, I would be in favor of marking all the pure functions `[[nodiscard]]`, as executing `getParameter` or `getAttribute` without assigning the return value should be an error. But I understand others may have different opinions, what do you think? Also tagging David here for another opinion @DavidFair .

**Comment by DavidFair** [comment]
> Personally, I would be in favor of marking most (there are some exceptions obviously) the pure functions `[[nodiscard]]`, as executing `getParameter` or `getAttribute` without assigning the return value should be an error. But I understand others may have different opinions, what do you think? Also tagging David here for another opinion @DavidFair .

I'm against this use, on the one hand it's a bug if you call a getter without taking a result. But on the other hand it's noise in the headers and someone on a code review or reading the code would see it unlike the `.empty()` example.
This also means test code which does

```
void test_no_throw_get_param(){
   m_testInstance.getParameter() // emits a compiler warning

  //---- so now we have to do 
  auto _ = m_testInstance.getParameter();
  UNUSED(_)  // suppress warning
```

Ditto for the throwing test case and so on

My "yard-stick" would be if there was something that was ambiguously named as to if it returned, say `addFunction` which returned a unique pointer or reference you had to hold to the new composite. From reading it you could naively assume the signature is `void addFunction(...)` when this would help steer them: `[[nodiscard]] std::unique_ptr<T> addFunction(...)`.

If you look at the STL on cppreference you'll see they [don't use it in many places](https://en.cppreference.com/w/cpp/language/attributes/nodiscard) and this software engineering [discussion says similar](https://softwareengineering.stackexchange.com/questions/363169/whats-the-reason-for-not-using-c17s-nodiscard-almost-everywhere-in-new-c)

That being said, this isn't me trying to put me people off using nodiscard, I suspect if were going to have sensible places to put them it will mostly likely be in the associated fitting headers - as the API has evolved with the project so it's got some nasty edges developers can catch themselves on (akin to `.empty()`)

**Comment by DavidFair** [comment]
Just for context, IIRC when this proposal was going through the STL committee the primary rational was to cut down on bugs with legacy interfaces:

``` c++

int myLegacyFileReader(char* out_ptr){
   return scanf(ptr);
}

char* user_input;
myLegacyFileReader(user_input); 
 // Bug - we called the legacy interface and now we've lost the number of chars read in, or worse if it returned 0 
// for EOL/error as we didn't store the return value
```

Which was unfortunately quite common with C system calls

**Comment by StephenSmith25** [comment]
I agree, it does create some noise in the header, but in that linked stackoverflow the top comment cites two use cases:

1. if a function has no effects beyond returning a certain result, i.e. is pure. If the result is not used, the call is certainly useless. On the other hand, discarding the result would not be incorrect.

2. if the return value must be checked, e.g. for a C-like interface that returns error codes instead of throwing. This is the primary use case. For idiomatic C++, that's going to be quite rare.

To me it seems like our pure getters in the function headers (Ifunction, CompositeFunction) fall under the first category. Is this not similar to how we get the build tools to stop us declaring unused variables, whereby this attribute would remove uncessary function calls. I guess theres no clear correct answer and opinions are divided, which is evident from google searches about this attribute.

**Comment by martyngigg** [comment]
I can see the argument that you certainly would get polluted headers if you basically mark every pure getter but it does feel like a call one of these types without capture is something you would want to explicitly say oh actually I really do mean that? I like the unused variable comparison as that's also a situation where it's technically harmless but it will probably confuse the reader.

File: qt/widgets/common/test/ConvolutionFunctionModelTest.h (hunk 1 above)
Comment (StephenSmith25) [reply qt/widgets/common/test/ConvolutionFunctionModelTest.h:364]: As far as I can tell it’s able to handle the new order. The ConvolutionFunctionModel is used in the ConvFunctionModel class which is subsequently used in the Indirect data analysis GUI and the tests for ConvFunctionModel and interface seem to behave as normal.

File: Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp [hunk 2]
```diff
@@ -401,8 +401,12 @@ void CrystalFieldFunction::buildAttributeNames() const {
   if (!m_attributeNames.empty()) {
     return;
   }
-  m_attributeNames = IFunction::getAttributeNames();
+  auto numAttributes = IFunction::nAttributes();
+  for (size_t i = 0; i < numAttributes; ++i) {
+    m_attributeNames.emplace_back(IFunction::attributeName(i));
```
[the remark this answers is not in the document]
Comment (StephenSmith25) [reply Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp:406]: I had to double check this, but I think the if statement above prevents this from being called more than once:

`if (!m_attributeNames.empty()) {return;}`

File: Framework/API/src/IFunction.cpp [hunk 3]
```diff
@@ -1342,9 +1342,9 @@ void IFunction::setAttributeValue(const std::string &attName,
  */
 void IFunction::setAttributeValue(const std::string &attName,
                                   const std::string &value) {
-  Attribute att = getAttribute(attName);
+  Attribute att = this->getAttribute(attName);
```
[the remark this answers is not in the document]
Comment (StephenSmith25) [reply Framework/API/src/IFunction.cpp:1345]: I think I was just trying a few things out at the start, these do not need to be qualified, so I've removed the `this` prefix.

File: qt/widgets/common/src/PropertyHandler.cpp [hunk 4]
```diff
@@ -982,8 +992,8 @@ void PropertyHandler::clearError(QtProperty *prop) {
  */
 Mantid::API::IFunction_sptr PropertyHandler::changeType(QtProperty *prop) {
   if (prop == m_type) {
-    // if (!m_parent) return m_browser->compositeFunction();// dont replace the
```
[the remark this answers is not in the document]
Comment (StephenSmith25) [reply qt/widgets/common/src/PropertyHandler.cpp:985]: Not that I can see, I've removed it now

File: Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp:406]: Yes quite right. Sorry. I was concentrating on the changed lines alone.

**Review by martyngigg — APPROVED** [review]
I'm happy with this now.

**Comment by StephenSmith25** [comment]
@mantidproject/gatekeepers can we get this merged

**Pull request merged**